### ISSUE TITLE
angular with tsconfig target ES2017 async/await will not work with zone.js

### ISSUE THREAD

**Issue by JiaLiPassion** [angular/zone.js#740]
this issue is similar with #715, if we use chrome v8 async/await and compile angular with tsconfig target 'ES2017', then typescript will not generate __awaiter code and use native async/await.
and the following logic will fail

```javascript
(click) = test();
const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
const test = async () => {
      console.log(Zone.current.name) // will output 'angular'
      await delay(100);
      console.log(Zone.current.name) // will output 'root'
    }
```

unlike typescript  transpiler, native async/await will first yield from test, and then call promise.then 
for continuation when await something. So Zone currentFrame will become root.
The sequence of above logic when call `await delay(100)` will look like 

1.  delay function return a ZoneAwarePromise, Zone.current is angular
2.  test function return native Promise which generate from chrome v8 by await, and test() execution is finished.
3. So Zone.currentZone is transite from angular->root
4. ZoneAwarePromise which generated from step1 was chained by called by `Promise.prototype.then`
5. delay timeout is executed, and ZoneAwarePromise resolved
6. the chained Promise is resolved , but the Zone.current is root.

**Comment by mhevery** [comment]
/FYI @domenic

The short answer is that it is not possible to intercept native promises in the VM. (Native promises is used by `async`/`await`) This is a known issue.

We have two approaches to this problem:
1) Try to standardize on Zone as part of the browser https://github.com/domenic/zones
2) Add hooks to the browser VM which would allow one to intercept native promises https://docs.google.com/document/d/1rda3yKGHimKIhg5YeoAmCOtyURgsbTH_qaYR79FELlk

Out of the two approaches I think second one (`setPromiseHook`) is more likely. 

This effort has been a bit on back-burner, but I will restart the discussions with @domenic.

**Comment by JiaLiPassion** [comment]
@mhevery , got it, I will also do some research about the v8 hooks. thank you!

**Comment by JiaLiPassion** [comment]
node.js begin to support `async_hooks`, 

https://github.com/nodejs/node-eps/blob/master/006-asynchooks-api.md
https://github.com/nodejs/node/pull/13000
https://github.com/nodejs/node/issues/13139#issuecomment-302945283

so I will try to implement zone.js with `async_hooks` in node.js, it should have better performance, and can handle native `async/await` issue in node.js.

@mhevery , do you think this is the correct direction? please review.

**Comment by mhevery** [comment]
Yes that is. I have worked with the people on async hook, and it should
work well with Zone. That means that we should be able to drop the Promise
polyfiil for node.

**Comment by JiaLiPassion** [comment]
@mhevery , got it, I will try to use `async-hook` to implement `Zone` in node.js and let you review after finish!

**Comment by amcdnl** [comment]
@JiaLiPassion - Awesome stuff! 
@mhevery - any idea when this could get merged up?

**Comment by JiaLiPassion** [comment]
@amcdnl, this is still under development, it will take some time, and I have to wait for nodejs to provide some additional information to finish this one.

**Comment by amcdnl** [comment]
Thanks for the update @JiaLiPassion ... I was able to temp work around it by transpiling my code from ES2017 to ES2016 which sucks but is ok for short term.

**Comment by enko** [comment]
Has there been any progress on this one? As far as my limited skill of perception can see, this is the main issue that prevents angulars change detection work properly with es2017's async/await.

ES2017 is realy helpfull for debugging as the compiled sourcecode by typescript/angular is not that different to the real source and so sourcemaps work much much better.

Thanks for your time, Tim.

**Comment by shellmann** [comment]
@JiaLiPassion can you say something about the current status?

**Comment by JiaLiPassion** [comment]
@enko, @shellmann, this issue is still pending because there is no proper way to patch `async/await` promise in `Javascript world`. In nodejs, it is possible to do the `patch` and I have made some POC to verify it. But in browser, it is still impossible to do that, so we will wait whether `zone` will pass `TC39` proposal or browser can open some `PromiseHook` like API to us.

So now, sorry this issue can not be resolved, so Angular can not work with `ES2017` currently.

**Comment by JiaLiPassion** [comment]
@enko, yes, it is, it has been in stage 0 for about two years...

**Comment by royalrover** [comment]
@JiaLiPassion  How about node.js now? Can  we resolve the question with async_hook? has examples? thx!

**Comment by JiaLiPassion** [comment]
@royalrover, yes, with `async_hooks`, this issue can be resolved, I am working on it in this PR #795.

**Comment by alamothe** [comment]
Just want to point out the "obvious" workaround for Node.js, which is to downgrade target to ES2015 😄 . Thanks for your great work!

**Comment by rvalimaki** [comment]
I get it that with current state of browsers this is not possible to fix by Angular team. However, it's a huge issue going forward with Angular lacking support for not-really-that-recent ES standards. As I get it, Google uses Angular a lot internally, there is some MS involvement as well and I get also that this issue requires "political" actions, not technical ones.

I hope people on Angular team understand how severe problems this little tiny issue #740 is going to cause moving forward, if those "political issues" to push browser support are not taken seriously by some more powerful entities within Google and maybe within MS as well. If those two giants were serious about the issue, it would be already fixed by now.

For now, the obvious workaround for the issue is just sitting on ES2015 target forever with horrendous generated async/await code, lack of optimizations and lack of proper debugging due to that generated code.

If we are just sitting and waiting "Zones for JavaScript" proposal to get anywhere from stage 0 of the TC39 process, after that was presented at the January 2016 TC39 meeting, we can be just waiting forever, without any hope about Angular ever supporting web standards going forward. You don't possess the power, but your bosses, or bosses of your bosses, or bosses of bosses of your bosses do possess the power, so please, push them hard!

**Comment by dfahlander** [comment]
I've followed this issue for some years now of curiosity and I think it's time to share my own solution to the problem, implemented in [Dexie](https://github.com/dfahlander/Dexie.js) version 2.0. It has been used for quite a while and works across all browsers keeping the zones between await calls. My first beta was released in October 2016, and the first stable 2.0 release in September 2017. Dexie has ~15k weekly downloads on npm and there are no show-stopping issues related to its zone system.

The reason for having a zone system in Dexie is to keep track of ongoing transactions. First version with the zone system built-in was release in 2014 (but then without support for async/await). This was before I knew about the concept of zones, so I thought it was my own invention at first, and called it PSD (Promise-Specific Data) as it only cares about promise-based async flows. The 2.0 version use something I call zone-echoing. It will enque micro-tasks that will re-enter the zone in coming micro-tasks. It can know when it's time to stop the zone echoing as long as the user sticks to its own Promises. A fallback will stop it by a limit. This works across all modern browsers and does not leak zones (except for MutationObserver subscriptions, whose events will derive the zone from the zone where the subscription was initialized - which wouldn't be a problem in angular/zone.js as it also patches those events).

The technique I use could be used in angular/zone.js but then every promise-returning function in the DOM would have to return a zone's own Promise (maybe it already does?) in order to invoke the zone echoing. There could be performance implications that would need to be considered. Especially as I detect native await calls by defining Promise.prototype.then using a getter instead of a value, and have to echo zones in order to support calls that awaits non-promises.

If interested, the implementation lies in Dexie repo at [src/helpers/promise.js](https://github.com/dfahlander/Dexie.js/blob/master/src/helpers/promise.js).

**Comment by JiaLiPassion** [comment]
@dfahlander, thank you for the post, I will definitely read it, and I will contact you if I have any questions.

**Comment by JiaLiPassion** [comment]
@dfahlander, thanks for your post again, I have sent you an email, please check it.

**Comment by JiaLiPassion** [comment]
@dfahlander, I think I have a basic understanding of your code, I tried the same way, it works in some cases, but not work in the case below.

```
async function test1() {
  console.log('before', Zone.current.name);
  await asyncFunction();
  console.log('after', Zone.current.name); // here is still not working
}

function test2() {
  test1();
}

Zone.current.fork({name: 'test'}).run(test2);
```

In this case, `test2` itself is not an `async` function, and inside it, it will call an `async` function, in this case, `zone.js` doesn’t know when `await of test1` begin, so `zone.js` can not keep the zone before `await`.

I think `Dexie` doesn't handle this too, because https://github.com/dfahlander/Dexie.js/blob/e956de3b74cf3ac5f1ed2fa2b97b53e4fe60a3e1/src/classes/dexie/transaction-helpers.ts#L57 the code here also require `scopeFunc` is an `AsyncFunction`, but if the `scopeFunc` is not `async`, and inside `scopeFunc` there is `await call`, it can't be traced.

Not sure my understanding is correct, please confirm, thanks!

**Comment by dfahlander** [comment]
That is true. It might not be the only solution. Maybe it would be better to always do `incrementExpectedAwaits()` and inspect whether returnValue is NativePromise, and if not, call `decrementExpectedAwaits()` directly. But I haven't thought it through...

**Comment by JiaLiPassion** [comment]
@dfahlander, thanks for the response, in this case, 
```
async function test1() {
  console.log('before', Zone.current.name);
  await asyncFunction();
  console.log('after', Zone.current.name); // here is still not working
}

function test2() {
  test1();
}

Zone.current.fork({name: 'test'}).run(test2);
```

the return value of function `test2` is `void`, so there is no way to check it is a `NativePromise or not`.
So we still need some `PromiseHook` to make all cases work!

**Comment by dfahlander** [comment]
True. Unless accompanied with a coding guideline of how to utilize zones together with native async await.

**Comment by JiaLiPassion** [comment]
@dfahlander, thanks for the clarification!

**Comment by rvalimaki** [comment]
Should this ticket have someone assigned on it? Now ES2019 (and ES2018) have been finalized, and Angular 8 is coming with support for producing both legacy (ES5) and modern (ES2015+) Javascript bundles. However, there is not much point of that, if modern ES versions ES2017, ES2018 and ES2019 are not supported anyway in Angular.

Is there any way to overcome this problem either by fixing it somehow or at least being able to use ES2017 or ES2018 target, but somehow use non-native async/await for time being?

**Comment by Eugeny** [comment]
Excuse my possibly stupid question, but why can't zone.js override the global `Promise` constructor to monkey-patch the `then` and `catch` methods instead of waiting for the native hooks?

**Comment by JiaLiPassion** [comment]
@rvalimaki, this is a very difficult issue to resolve, I am still trying to find out a walk around.
@Eugeny, current `zone.js` already monkey-patch `Promise`, but `async/await` is different, it does not use `javascript Promise`, it always use `Native Promise`.

**Comment by Eugeny** [comment]
@JiaLiPassion thanks for clarifying

**Comment by rvalimaki** [comment]
@JiaLiPassion Good to hear that there is still ongoing effort to fix this. 

Has Google Angular team (@mhevery et. al) contacted  Google Chrome/Chromium team about the issue to provide some kind of hooks for native promises? To my knowledge Google is using Angular in about 600 projects, and I guess it should be quite high on their priority to keep Angular up with modern web standards.

To my understanding the proposal for implementing Zones in EcmaScript is dead, as it's still in "stage 0" after more than three years: https://github.com/domenic/zones

If we didn't have so large code base already, I would explore about the possibility to use Angular without Zones, though then we would need to call "app.tick()" manually every time after updates which would be really a major PITA for sure.

**Comment by JiaLiPassion** [comment]
@rvalimaki. 
To summarize. 
1. `Zone` proposal is not updated for a long time.
2. We may expect `Chrome` team release `PromiseHook` in Javascript like `nodejs`, but it is a very very long shot.
3. With `Angular Ivy` in the future, there will be a `non zone` mode, and user need to trigger change detection themselves.
